main/app.c

```diff
@@ -2306,16 +2623,194 @@ int ast_app_group_update(struct ast_channel *old, struct ast_channel *new)
 	return 0;
 }
 
+int ast_app_group_add_meta_ifneeded(const char *group, const char *category)
+{
+	struct ast_group_meta *gmi = NULL; /*!< Group metadatas */
+
+	AST_RWLIST_WRLOCK(&groups_meta);
+	AST_RWLIST_TRAVERSE_SAFE_BEGIN(&groups_meta, gmi, group_meta_list) {
+		if (!strcasecmp(gmi->group, group) && !strcasecmp(gmi->category, category)) {
+			break; /* We only have one list item per group@category */
+		}
+	}
+	AST_RWLIST_TRAVERSE_SAFE_END;
+
+	if (!gmi) {
+		if (!(gmi = ast_calloc(1, sizeof(struct ast_group_meta)))) {
+		    AST_RWLIST_UNLOCK(&groups_meta);
+		    return -1;
+		}
+
+		ast_copy_string(gmi->group, group, MAX_GROUP_LEN);
+
+		if (!ast_strlen_zero(category)) {
+			ast_copy_string(gmi->category, category, MAX_CATEGORY_LEN);
+		}
+
+		AST_RWLIST_INSERT_TAIL(&groups_meta, gmi, group_meta_list);
+
+		manager_event(EVENT_FLAG_DIALPLAN, "GroupCreate",
+			"Category: %s\r\n"
+			"Group: %s\r\n",
+			category, group);
+	}
+
+	gmi->num_channels++;
+
+	AST_RWLIST_UNLOCK(&groups_meta);
+
+	return 1;
+}
+
+/* Remove all channels from the given group */
+int ast_app_group_remove_all_channels(const char *group, const char *category)
+{
+	struct ast_group_info *gi = NULL;
+
+	int group_meta_found = 0;
+
+	if (!group) {
+		ast_log(LOG_WARNING, "Group Remove: At the minimum, group must be non-NULL\n");
```

Comment:
If this is for the user, this should be rephrased to something like a group must be provided rather than non-NULL, since the dialplan doesn't have the concept of NULL

Reply:
This error is for a developer.  ast_app_group_remove_all_channels would never be called from user-code with NULL

The only usage of this is via AMI manager_group_remove
```
static int manager_group_remove(struct mansession *s, const struct message *m)
{
        const char *group    = astman_get_header(m, "Group");
        const char *category = astman_get_header(m, "Category");

        if (ast_app_group_remove_all_channels(group, category) == 0) {
```



group/category will never be null here.